pkg/client/session.go

```diff
@@ -0,0 +1,79 @@
+package client
+
+import (
+	"context"
+	"fmt"
+	"github.com/codenotary/immudb/pkg/api/schema"
+	"github.com/codenotary/immudb/pkg/client/cache"
+	"github.com/codenotary/immudb/pkg/client/errors"
+	"github.com/codenotary/immudb/pkg/client/heartbeater"
+	"github.com/codenotary/immudb/pkg/client/state"
+	"github.com/codenotary/immudb/pkg/signer"
+	"github.com/codenotary/immudb/pkg/stream"
+	"github.com/golang/protobuf/ptypes/empty"
+	"google.golang.org/grpc"
+)
+
+func (c *immuClient) OpenSession(ctx context.Context, user []byte, pass []byte, database string) (serverUUID string, sessionID string, err error) {
+	c.Options.DialOptions = c.SetupDialOptions(c.Options)
+	if db, e := c.Tkns.GetDatabase(); e == nil && len(db) > 0 {
+		c.Options.CurrentDatabase = db
+	}
+
+	if c.Options.ServerSigningPubKey != "" {
+		pk, e := signer.ParsePublicKeyFile(c.Options.ServerSigningPubKey)
+		if e != nil {
+			return "", "", e
+		}
+		c.WithServerSigningPubKey(pk)
+	}
+
+	if c.Options.StreamChunkSize < stream.MinChunkSize {
+		return "", "", errors.New(stream.ErrChunkTooSmall).WithCode(errors.CodInvalidParameterValue)
+	}
+
+	if c.clientConn, err = grpc.Dial(c.Options.Bind(), c.Options.DialOptions...); err != nil {
+		return "", "", err
+	}
+
+	serviceClient := schema.NewImmuServiceClient(c.clientConn)
+	// todo remove WithServiceClient
+	c.WithServiceClient(serviceClient)
+
+	resp, err := c.ServiceClient.OpenSession(ctx, &schema.OpenSessionRequest{
+		User:         user,
+		Password:     pass,
+		DatabaseName: database,
+	})
+	if err != nil {
+		return "", "", errors.FromError(err)
+	}
+
+	c.SessionID = resp.GetSessionID()
+
+	if err = heartbeater.NewHeartBeater(resp.GetSessionID(), c.ServiceClient).KeepAlive(); err != nil {
```

Comment:
How is heartbeater stopped? We don't even store pointer to it anywhere.

Reply:
I will stop it on Close session